TTL for cleaning up Pods after they finish

alaypatel07 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**: [enhancements/issues/592](https://github.com/kubernetes/enhancements/issues/592)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
For [enhancements/issues/592](https://github.com/kubernetes/enhancements/issues/592)

**Special notes for your reviewer**: @kubernetes/sig-apps-pr-reviews

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Add a TTL mechanism to clean up Pods after they finish.
```

alaypatel07 committed:
Modify ttlafterfinished_controller, make it generic for multiple gvks

alaypatel07 committed:
Add tests for generic ttlafterfinished_controller

alaypatel07 commented:
/assign @janetkuo

janetkuo commented:
/ok-to-test

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -20,92 +20,107 @@ import (
 
-// New creates an instance of Controller
-func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *Controller {
+// New creates an instance o Controller
```
typo?

alaypatel07 replied:
yeah, I'll fix it. Thanks for spotting!

janetkuo commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -481,9 +482,15 @@ func startTTLAfterFinishedController(ctx ControllerContext) (http.Handler, bool,
+		ctx.InformerFactory,
+		schema.GroupVersionResource{Version: "v1", Resource: "pods"},
+		dynamic.NewForConfigOrDie(ctx.ClientBuilder.ConfigOrDie("ttl-pods-after-finished-controller")),
+		ctx.ClientBuilder.ClientOrDie("ttl-pods-after-finished-controller")).Run(int(ctx.ComponentConfig.TTLAfterFinishedController.ConcurrentTTLSyncs), ctx.Stop)
```
We only need to start a single instance of TTL controller that handles both Jobs and Pods.

tnozicka replied:
I agree, start one instance of the controller, 2 informers, 1 queue and then switch based on type where the flow would differ. We can go to dynamic informers and unstructured objects checking if we determine we need it for CRDs later.

janetkuo commented:
/cc @tnozicka

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -20,92 +20,107 @@ import (
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.rListerSynced) {
 		return
 	}
 
```
We could just run a dedicated worker for Pods, and a dedicated queue/informer/lister for Pods too. 

We could use something more generic like what [garbage collector](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go) uses, but that seems to be premature optimization. We can move to use that if we're supporting custom resources.

alaypatel07 replied:
@janetkuo Thanks. I’ll refactor the code with dedicated queue/informer/lister for pods and jobs

Out of curiosity, can I ask why not use two controllers for jobs and pods if we are using dedicated queues/informers/lister(which I think is the crux of a controller) anyways? Am I missing anything?

xmudrii commented:
@alaypatel07 @janetkuo @tnozicka Hello! I'm the bug triage lead for the 1.16 release cycle and considering this PR is tagged for 1.16, I'd like to check its status. The code freeze is starting on August 29th EOD PST (about 5 days from now) and once the code freeze starts, only release-blocking PRs will be considered to be merged. Is this PR still planned for the 1.16 release?

alaypatel07 committed:
Refactor ttlafterfinished_controller to have multiple listers

This includes a separate lister for Jobs and Pods in the same
Controller. We will use unstructured later CRDs need to be
supported.

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -179,93 +195,109 @@ func (tc *Controller) handleErr(err error, key interface{}) {
-func needsCleanup(j *batch.Job) bool {
+// needsCleanup checks whether a resource has finished and has a TTL set.
+func needsCleanup(obj interface{}) bool {
+	if job, ok := obj.(*batch.Job); ok {
```
do a switch and hard error / panic if none of the types is matched

janetkuo commented:
@xmudrii yes this PR is still planned for 1.16.

ipochi commented:
@tnozicka @janetkuo 

@alaypatel07 Hey I created a PR #82082 for the same
Would you be able to get your PR ready for merging. I will get reviewed and close without merging  if you can.

alaypatel07 commented:
@ipochi Thanks for the heads up. I am trying to finish it up today. Hopefully, that would be enough time for merging.

liggitt commented:
/milestone v1.17

this has missed code freeze

alejandrox1 commented:
@alaypatel07 @janetkuo @enisoc @tnozicka any updates on this?
/remove-area e2e-test-framework
/remove-sig testing

/remove-area cloudprovider
/remove-area code-generation
/remove-area conformance
/remove-area dependency
/remove-area ipvs
/remove-area kubeadm
/remove-area kubectl
/remove-area release-eng
/remove-area kubelet
/remove-area test
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig node
/remove-sig release
/remove-sig scheduling
/remove-sig storage

alejandrox1 commented:
/cc

josiahbjorgaard commented:
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged by then? If not, can we move it out of the milestone?

alejandrox1 commented:
/milestone clear

alaypatel07 commented:
closing because another PR is tracking similar work
